### PR TITLE
feat: vercel edge sdk

### PR DESCRIPTION
**Pull request by monsagri**
[sc-194414]

This adds an Edge SDK for Vercel to be used with Vercel Edge Configs. 

I've added some basic testing, and using the built package in a vercel project using npm link works as expected.

### CODE REVIEW

File: packages/sdk/vercel/src/createLdClient/createOptions.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import packageJson from '../../package.json';
+import createFeatureStore from './createFeatureStore';
+
+type SupportedLDOptions = Pick<LDOptions, 'logger' | 'featureStore'>;
+const allowedOptions = ['logger', 'featureStore'];
+
+const defaults = {
+  stream: false,
+  // TODO: Investigate if we can actually send events
+  sendEvents: false,
```
Comment (monsagri) [inline packages/sdk/vercel/src/createLdClient/createOptions.ts:12]: Leaving this for a follow up PR as it's a nice to have

File: packages/sdk/vercel/src/platform/crypto/index.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  createHash(algorithm: SupportedHashAlgorithm): Hasher {
+    return new CryptoJSHasher(algorithm);
+  }
+
+  createHmac(algorithm: SupportedHashAlgorithm, key: string): Hmac {
+    return new CryptoJSHmac(algorithm, key);
+  }
+
+  randomUUID(): string {
+    return uuidv4();
```
Comment (monsagri) [inline packages/sdk/vercel/src/platform/crypto/index.ts:25]: Feel like I'm doing something wrong here, see todo above

Comment (kinyoklion) [reply packages/sdk/vercel/src/platform/crypto/index.ts:25]: Does crypto.randomUUID work in vercel?

https://vercel.com/templates/next.js/edge-functions-crypto

**Comment by kinyoklion** [comment]
We figured we would need a shared folder for some potentially common edge provider items, but it is hard to say for sure until you make 2. :-)

File: packages/sdk/vercel/src/createLdClient/createFeatureStore.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  LDFeatureStore,
+  LDFeatureStoreDataStorage,
+  LDFeatureStoreItem,
+  LDFeatureStoreKindData,
+} from '@launchdarkly/js-server-sdk-common';
+import noop from '../utils/noop';
+
+const createFeatureStore = (edgeConfig: EdgeConfigClient, sdkKey: string, logger: LDLogger) => {
+  const key = `LD-Env-${sdkKey}`;
+  const store: LDFeatureStore = {
```
Comment (kinyoklion) [inline packages/sdk/vercel/src/createLdClient/createFeatureStore.ts:14]: Could you implement the feature store as a class? `class VercelFeatureStore implements LDFeatureStore`?

File: packages/sdk/vercel/src/createLdClient/createFeatureStore.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    get(
+      kind: DataKind,
+      flagKey: string,
+      callback: (res: LDFeatureStoreItem | null) => void = noop
+    ): void {
+      logger.debug(`Requesting ${flagKey} from ${key}`);
+      edgeConfig
+        .get(key)
+        .then((i) => {
```
Comment (kinyoklion) [inline packages/sdk/vercel/src/createLdClient/createFeatureStore.ts:23]: Potentially you could just call an async function inside these, passing it the callback, and then not await it. So that the code could be structured using async/await.

Potentially we could add a more standard async feature of LDFeatureStore. I had just kept the interface largely the same for simplicity of porting existing stores.

I generally find async code much easier to follow and maintain. But this is just a thought/suggestion, not something that has to be changed.

File: packages/sdk/vercel/src/createLdClient/createOptions.ts (hunk 1 above)
Comment (kinyoklion) [reply packages/sdk/vercel/src/createLdClient/createOptions.ts:12]: If vercel has fetch, then it should be possible. Not sure on the value, because I have not really been in the loop how we use these edge SDKs and why it is fine for them to not send events.

File: packages/sdk/vercel/src/platform/crypto/index.ts (hunk 2 above)
Comment (monsagri) [reply packages/sdk/vercel/src/platform/crypto/index.ts:25]: 👍 thanks!

File: packages/sdk/vercel/src/createLdClient/createOptions.ts (hunk 1 above)
Comment (InTheCloudDan) [reply packages/sdk/vercel/src/createLdClient/createOptions.ts:12]: I left a slack comment for @yusinto in regards to Cloudflare. When we go to enable `events` we want to also add in the `track` call. I'm not sure if it's waiting until `sendEvents` is usable to add that in.

Comment (kinyoklion) [reply packages/sdk/vercel/src/createLdClient/createOptions.ts:12]: Yeah, we probably wouldn't want to add it before it actually sends events.

**Comment by yusinto** [comment]
Please take a look at #83 which has the common code for cloudflare and vercel refactored to packages/shared/sdk-server-edge.

**Comment by monsagri** [comment]
> Please take a look at #83 which has the common code for cloudflare and vercel refactored to packages/shared/sdk-server-edge.

Got it thanks! Actually had a scroll through already and will be making sure to get this updated today/tomorrow latest

Thanks for doing all the heavy lifting with the shared stuff!

**Comment by yusinto** [comment]
> Thanks for doing all the heavy lifting with the shared stuff!

No worries! Thank you for doing the Vercel sdk. I have merged #83 now so you'll be able to use it from main which makes it easier.

**Comment by yusinto** [comment]
Please take a look at #87 for producing cjs and esm bundles for the cloudflare sdk. TLDR your build command should use `scripts/build-package.sh` which will do the job. The [relevant code](https://github.com/launchdarkly/js-core/blob/5c45173edc67a9e4b0134a90320d337aef5f8de7/packages/sdk/cloudflare/package.json#L12) are:

```json
  "type": "module",
  "exports": {
    "require": "./dist/cjs/src/index.js",
    "import": "./dist/esm/src/index.js"
  },
  "main": "./dist/cjs/src/index.js",
  "types": "./dist/cjs/src/index.d.ts",
  "files": [
    "/dist"
  ],
  "scripts": {
    "build": "../../../scripts/build-package.sh",
```

**Comment by yusinto** [comment]
The Cloudflare SDK 0.0.1 has been published. All code are in main now so you can safely follow the cloudflare sdk for vercel. Let me know if everything makes sense. Good luck!

File: packages/sdk/vercel/src/createLdClient/createFeatureStore.ts (hunk 4 above)
Comment (monsagri) [reply packages/sdk/vercel/src/createLdClient/createFeatureStore.ts]: Not sure if I totally followed what you meant by not awaiting within the implementation here, but the change is an easy to revert commit and seems to work as expected.

**Comment by monsagri** [comment]
I've made the changes that we talked about and adapted everything to use the new common modules where possible. 
Missing some tests against the main implementation for now using something like miniflare to mock the edge config, but think that can come as a follow up?

File: packages/sdk/vercel/package.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  "keywords": [
+    "launchdarkly",
+    "vercel",
+    "edge"
+  ],
+  "type": "module",
+  "exports": {
+    "require": "./dist/cjs/src/index.js",
+    "import": "./dist/esm/src/index.js",
+    "types": "./dist/esm/src/index.d.ts"
```
Comment (yusinto) [inline packages/sdk/vercel/package.json:15]: `types` can be omitted because typescript by default will look for definition files next to index.js.

File: packages/sdk/vercel/package.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  "main": "./dist/cjs/src/index.js",
+  "types": "./dist/cjs/src/index.d.ts",
+  "files": [
+    "/dist"
+  ],
+  "scripts": {
+    "doc": "../../../scripts/build-doc.sh .",
+    "build": "../../../scripts/build-package.sh",
+    "tsw": "yarn tsc --watch",
+    "start": "yarn tsw",
```
Comment (yusinto) [inline packages/sdk/vercel/package.json:26]: You may want to clean the `dist` folder first before `yarn start`

File: packages/sdk/vercel/package.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    "lint": "eslint . --ext .ts",
+    "prettier": "prettier --write '**/*.@(js|ts|tsx|json|css)'",
+    "test": "jest --ci --runInBand",
+    "check": "yarn prettier && yarn lint && yarn build && yarn test && yarn doc"
+  },
+  "dependencies": {
+    "@launchdarkly/js-server-sdk-common": "^0.2.0",
+    "@launchdarkly/js-server-sdk-common-edge": "0.0.2",
+    "crypto-js": "^4.1.1",
+    "uuid": "^9.0.0"
```
Comment (yusinto) [inline packages/sdk/vercel/package.json:36]: `@launchdarkly/js-server-sdk-common` and `uuid` should be omitted because they are unused.

File: packages/sdk/vercel/package.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  },
+  "devDependencies": {
+    "@types/crypto-js": "^4.1.1",
+    "@types/jest": "^29.5.0",
+    "@types/uuid": "^9.0.1",
```
Comment (yusinto) [inline packages/sdk/vercel/package.json:41]: `@types/uuid` should be omitted because uuid is unused.

File: packages/sdk/vercel/package.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    "/dist"
+  ],
+  "scripts": {
+    "doc": "../../../scripts/build-doc.sh .",
+    "build": "../../../scripts/build-package.sh",
+    "tsw": "yarn tsc --watch",
+    "start": "yarn tsw",
+    "lint": "eslint . --ext .ts",
+    "prettier": "prettier --write '**/*.@(js|ts|tsx|json|css)'",
+    "test": "jest --ci --runInBand",
```
Comment (yusinto) [inline packages/sdk/vercel/package.json:29]: Jest with coverage reveals that `index.ts` and `createPlatformInfo.ts` are completely untested. Are you able to add unit tests for those please?

```json
"test": "jest --ci --runInBand --coverage",
```

Screenshot 2023-04-21 at 10 14 44 PM

File: package.json [hunk 10]
```diff
@@ -6,7 +6,8 @@
     "packages/shared/sdk-server-edge",
     "packages/sdk/server-node",
     "packages/sdk/cloudflare",
-    "packages/sdk/cloudflare/example"
+    "packages/sdk/cloudflare/example",
+    "packages/sdk/vercel"
```
Comment (yusinto) [inline package.json:10]: If you have time, it will be good to add an example app to demo the basic usage of the vercel sdk.

File: scripts/build-package.sh [hunk 11]
```diff
@@ -23,8 +23,8 @@ ESM_PACKAGE_JSON=$( jq -n \
                   --arg type "module" \
                   '{ name: $name, version: $version, type: $type }' )
 
-tsc --module commonjs --outDir dist/cjs/
+tsc --module commonjs --declaration --outDir dist/cjs/ 
 echo "$CJS_PACKAGE_JSON" > dist/cjs/package.json
 
-tsc --module es2022 --outDir dist/esm/
+tsc --module es2022 --declaration --outDir dist/esm/
```
Comment (yusinto) [inline scripts/build-package.sh:29]: Is `--declaration` needed? I tested without and definition files are generated so I think they are unnecessary.

**Review by yusinto — CHANGES_REQUESTED** [review]
Nice work! I requested some changes but nothing serious.

You will also need to modify `release-please-config.json` for the initial publishing of this package. See [this](https://www.typescriptlang.org/tsconfig/#declaration).

File: packages/sdk/vercel/package.json (hunk 5 above)
Comment (monsagri) [reply packages/sdk/vercel/package.json]: I thought I needed them from testing the build output via npm link, but looks like I was wrong

File: packages/sdk/vercel/package.json (hunk 9 above)
Comment (monsagri) [reply packages/sdk/vercel/package.json]: Yep! 
Good shout, I've added them now and all files look better

File: package.json (hunk 10 above)
Comment (monsagri) [reply package.json:10]: Any chance we can do that as a follow up ticket? 

I'd like to get this merged for now so we can finalise the other work and get ready to release - we'll also need to set up a demo app for the vercel store page, so can bundle that work at that point.

**Review by yusinto — APPROVED** [review]
